File: apollo-runtime/src/jvmMain/java/com/apollographql/apollo/ApolloQueryWatcher.java
package com.apollographql.apollo;

import com.apollographql.apollo.api.Operation;
import com.apollographql.apollo.fetcher.ResponseFetcher;
import com.apollographql.apollo.internal.util.Cancelable;

import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

public interface ApolloQueryWatcher<T> extends Cancelable {

  ApolloQueryWatcher<T> enqueueAndWatch(@Nullable ApolloCall.Callback<T> callback);

  /**
   * @param fetcher The {@link ResponseFetcher} to use when the call is refetched due to a field changing in the
   * cache.
   */
  @NotNull ApolloQueryWatcher<T> refetchResponseFetcher(@NotNull ResponseFetcher fetcher);

  /**
   * Returns GraphQL watched operation.
   *
   * @return {@link Operation}
   */
  @NotNull Operation operation();

  /**
   * Re-fetches watched GraphQL query.
   */
  void refetch();

  /**
   * Cancels this {@link ApolloQueryWatcher}. The {@link com.apollographql.apollo.ApolloCall.Callback}
   * will be disposed, and will receive no more events. Any active operations will attempt to abort and
   * release resources, if possible.
   */
  @Override void cancel();

  /**
   * Creates a new, identical call to this one which can be enqueued or executed even if this call has already been.
   *
   * @return The cloned ApolloQueryWatcher object.
   */
  @NotNull ApolloQueryWatcher<T> clone();
}

File: deprecated-apollo-runtime/src/main/java/com/apollographql/apollo3/ApolloQueryWatcher.kt
package com.apollographql.apollo3

import com.apollographql.apollo3.api.Operation
import com.apollographql.apollo3.fetcher.ResponseFetcher
import com.apollographql.apollo3.internal.util.Cancelable

interface ApolloQueryWatcher<D : Operation.Data> : Cancelable {
  fun enqueueAndWatch(callback: ApolloCall.Callback<D>): ApolloQueryWatcher<D>

  /**
   * @param fetcher The [ResponseFetcher] to use when the call is refetched due to a field changing in the
   * cache.
   */
  fun refetchResponseFetcher(fetcher: ResponseFetcher): ApolloQueryWatcher<D>

  /**
   * Returns GraphQL watched operation.
   *
   * @return [Operation]
   */
  fun operation(): Operation<*>

  /**
   * Re-fetches watched GraphQL query.
   */
  fun refetch()

  /**
   * Cancels this [ApolloQueryWatcher]. The [com.apollographql.apollo3.ApolloCall.Callback]
   * will be disposed, and will receive no more events. Any active operations will attempt to abort and
   * release resources, if possible.
   */
  override fun cancel()

  /**
   * Creates a new, identical call to this one which can be enqueued or executed even if this call has already been.
   *
   * @return The cloned ApolloQueryWatcher object.
   */
  fun clone(): ApolloQueryWatcher<D>
}